src/wp-includes/user.php

```diff
@@ -3629,53 +3629,101 @@ function _wp_get_current_user() {
 function send_confirmation_on_profile_email() {
 	global $errors;
 
-	$current_user = wp_get_current_user();
 	if ( ! is_object( $errors ) ) {
 		$errors = new WP_Error();
 	}
 
-	if ( $current_user->ID != $_POST['user_id'] ) {
-		return false;
-	}
+	$user       = get_user_by( 'id', $_POST['user_id'] );
+	$email_sent = send_user_email_change_confirmation_email( $user, $_POST['email'] );
 
-	if ( $current_user->user_email != $_POST['email'] ) {
-		if ( ! is_email( $_POST['email'] ) ) {
+	if ( is_wp_error( $email_sent ) ) {
+		// WP_Error::copy_errors() with the addition of adding data.
+		foreach ( $email_sent->get_error_codes() as $code ) {
 			$errors->add(
-				'user_email',
-				__( '<strong>Error:</strong> The email address is not correct.' ),
+				$code,
+				$email_sent->get_error_message( $code ),
 				array(
 					'form-field' => 'email',
 				)
 			);
-
-			return;
 		}
+	}
 
-		if ( email_exists( $_POST['email'] ) ) {
-			$errors->add(
-				'user_email',
-				__( '<strong>Error:</strong> The email address is already used.' ),
-				array(
-					'form-field' => 'email',
-				)
-			);
-			delete_user_meta( $current_user->ID, '_new_email' );
+	if ( true === $email_sent ) {
+		$_POST['email'] = $user->user_email;
+	}
+}
 
-			return;
-		}
+/**
+ * Send the 'confirm your email' email.
+ *
+ * @since x.x
+ *
+ * @param WP_User $user  The user to act upon.
+ * @param string  $email The new email address.
+ * @return null|true|WP_Error true if email sent, WP_Error on error, and null otherwise.
+ */
+function send_user_email_change_confirmation_email( $user, $email ) {
+	if ( $user->user_email == $email ) {
+		return;
+	}
 
-		$hash           = md5( $_POST['email'] . time() . wp_rand() );
-		$new_user_email = array(
-			'hash'     => $hash,
-			'newemail' => $_POST['email'],
+	if ( ! is_email( $email ) ) {
+		return new WP_Error(
+			'user_email',
+			__( '<strong>Error:</strong> The email address is not correct.' )
+		);
+	}
+
+	if ( email_exists( $email ) ) {
+		delete_user_meta( $user->ID, '_new_email' );
+
+		return new WP_Error(
+			'user_email',
+			__( '<strong>Error:</strong> The email address is already used.' )
 		);
-		update_user_meta( $current_user->ID, '_new_email', $new_user_email );
+	}
+
+	// The email is only sent if a user changes their own email.
+	$should_send_email_for_change = ( get_current_user_id() === $user->ID );
+
+	/**
+	 * Filters whether a 'confirm your email address' email should be sent.
+	 * If false is returned, the change is made immediately.
+	 *
+	 * @since x.x
+	 *
+	 * @param bool    $should_send_email_for_change Whether to use an email confirmation.
+	 * @param WP_User $user                         The user having their email changed.
+	 * @param string  $email                        The new email address.
+	 */
+	$should_send_email_for_change = apply_filters( 'should_send_email_for_email_change', $should_send_email_for_change, $user, $email );
```

Comment:
This filter can be used instead of `remove_action( 'personal_options_update', 'send_confirmation_on_profile_email' );`.

```
// Skip email confirmation for user email changes.
add_filter( 'should_send_email_for_email_change', '__return_false' );
```